src/Redist/Microsoft.ML.TensorFlow.Redist/Microsoft.ML.Tensorflow.Redist.proj

```diff
@@ -0,0 +1,133 @@
+    <MakeDir Directories="$(IntermediateOutputPath)" />
+    <ItemGroup>
+      <!-- DownloadFilesFromUrl requires Url metadata is set and ignores the identity of the item -->
+      <_downloadFiles Include="@(TensorFlowArchive)" Url="%(Identity)" DestinationFile="%(DownloadFile)" />
+    </ItemGroup>
+    <DownloadFilesFromUrl Items="@(_downloadFiles)">
+      <Output TaskParameter="FilesCreated" ItemName="FileWrites" />
+    </DownloadFilesFromUrl>
+  </Target>
+  
+  <UsingTask TaskName="GenerateChecksums" AssemblyFile="$(ToolsDir)Microsoft.DotNet.Build.Tasks.dll"/>
+  <UsingTask TaskName="ZipFileExtractToDirectory" AssemblyFile="$(ToolsDir)Microsoft.DotNet.Build.Tasks.dll"/>
+  <Target Name="ValidateAndExtractArchives" 
+          DependsOnTargets="DownloadArchives"
+          Inputs="@(TensorFlowArchive->'%(DownloadFile)')"
+          Outputs="@(TensorFlowArchive->'%(ExtractSemaphore)')">
+    <!-- GenerateChecksums writes a sha checksum to the file specified by DestinationPath metadata -->
+    <ItemGroup>
+      <_filesToCheckSum Include="@(TensorFlowArchive->'%(DownloadFile)')" DestinationPath="%(DownloadShaFile)" />
+    </ItemGroup>
+    <GenerateChecksums Items="@(_filesToCheckSum)" />
+    <ItemGroup>
+      <FileWrites Include="@(TensorFlowArchive->'%(DownloadShaFile)')" />
+    </ItemGroup>
+
+    <!-- If specified we'll update the checked in SHAs with the downloaded ones. -->
+    <Copy Condition="'$(UpdateSHA)' == 'true'"
+          SourceFiles="@(TensorFlowArchive->'%(DownloadShaFile)')"
+          DestinationFiles="@(TensorFlowArchive->'%(LocalShaFile)')" />
+
+    <Error Condition="!Exists('%(TensorFlowArchive.LocalShaFile)')" Text="SHA file '%(TensorFlowArchive.LocalShaFile)' does not exist.  Build with /p:UpdateSHA=true to save it." />
+
+    <!-- Read in the SHAs as metadata on the archive items -->
+    <ItemGroup>
+      <TensorFlowArchive>
+        <LocalSha>$([System.IO.File]::ReadAllText('%(LocalShaFile)'))</LocalSha>
+        <DownloadSha>$([System.IO.File]::ReadAllText('%(DownloadShaFile)'))</DownloadSha>
+      </TensorFlowArchive>
+    </ItemGroup>
+    <!-- Validate that the downloaded SHAs match the expected checked in SHAs -->
+    <Error Condition="'%(TensorFlowArchive.LocalSha)' != '%(TensorFlowArchive.DownloadSha)'" Text="Downloaded file '%(TensorFlowArchive.DownloadFile)' has unexpected SHA.%0A  expected: %(_downloadedTensorFlowArchive.LocalSha)%0A  actual: %(_downloadedTensorFlowArchive.DownloadSha)%0ABuild with /p:UpdateSHA=true if you intentionally changed the URL and wish to update the SHAs, otherwise this could indicate an incomplete download or intercerpted URL and should be examined." />
+
+    <!-- The archives are valid, lets extract them, ensuring an empty directory -->
+    <RemoveDir Directories="@(TensorFlowArchive->'%(ExtractDirectory)')" />
+    <MakeDir Directories="@(TensorFlowArchive->'%(ExtractDirectory)')" />
+    <ZipFileExtractToDirectory Condition="'%(FileExtension)' == '.zip'"
```

Comment:
(nit) To be complete, shouldn't this be `'%(TensorFlowArchive.FileExtension)'`?  Same for the `.tar.gz` check below.